prow/plugins/hold/hold.go

```diff
@@ -0,0 +1,116 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package hold contains a plugin which will allow users to label their
+// own pull requests as not ready or ready for merge. The submit queue
+// will honor the label to ensure pull requests do not merge when it is
+// applied.
+package hold
+
+import (
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "hold"
+
+var (
+	label         = "do-not-merge/hold"
+	labelRe       = regexp.MustCompile(`(?mi)^/hold\s*$`)
+	labelCancelRe = regexp.MustCompile(`(?mi)^/hold cancel\s*$`)
+)
+
+type event struct {
+	org           string
+	repo          string
+	number        int
+	prAuthor      string
+	commentAuthor string
+	body          string
+	htmlurl       string
+	hasLabel      bool
+}
+
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	CreateComment(owner, repo string, number int, comment string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+}
+
+func handleIssueComment(pc plugins.PluginClient, ic github.IssueCommentEvent) error {
+	// Only consider open PRs.
+	if !ic.Issue.IsPullRequest() || ic.Issue.State != "open" || ic.Action != github.IssueCommentActionCreated {
+		return nil
+	}
+
+	e := &event{
+		org:           ic.Repo.Owner.Login,
+		repo:          ic.Repo.Name,
+		number:        ic.Issue.Number,
+		prAuthor:      ic.Issue.User.Login,
+		commentAuthor: ic.Comment.User.Login,
+		body:          ic.Comment.Body,
+		hasLabel:      ic.Issue.HasLabel(label),
+		htmlurl:       ic.Comment.HTMLURL,
+	}
+	return handle(pc.GitHubClient, pc.Logger, e)
+}
+
+const warning = "only PR authors can issue or cancel holds."
+
+// handle drives the pull request to the desired state. If an author adds
+// a /hold directive, we want to add a label if one does not already exist.
+// If they add /hold cancel, we want to remove the label if it exists. If
+// a non-author adds either, we want to leave a comment letting them know
+// they can't hold PRs unless they are the author.
+func handle(gc githubClient, log *logrus.Entry, e *event) error {
+	needsLabel := false
+	if labelRe.MatchString(e.body) {
+		needsLabel = true
+	} else if labelCancelRe.MatchString(e.body) {
+		needsLabel = false
+	} else {
+		return nil
+	}
+
+	if e.commentAuthor != e.prAuthor {
```

Comment:
I prefer to always start with an assumption that our community is composed of collaborative people with good intentions, and make system as flexible and collaborative as possible.

If we find an individual acting poorly my instinct would be to start by attempting to inform that person about our community's social norms. 

If we discover that our assumption is wrong and our community is riddled with trolls then that is a great time to lock down useful functionality.

Reply:
Ah, ok! I misunderstood the requirements you put out on Slack! I'll cut this bit out -- it will make the flow simpler as well!